File: CS/GroupSorting/GroupSummaryOptions.cs
// Developer Express Code Central Example:
// How to apply sorting by Summary along with sorting by the Column.
// 
// This example shows how to sort group summary items along with the column that
// the summary is aligned with.
// To add this functionality, you can use the
// proposed helper class (GroupSummaryOptions) in your solution. Just set the
// attached property of the Grid to "True".
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E4611

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using DevExpress.Xpf.Grid;

namespace GridControlOptions {
    public class GroupSummaryOptions {
        #region Attached Properties

        #region SortGroupSummaryWithColumn
        public static bool GetSortGroupSummaryWithColumn(DependencyObject obj) {
            return (bool)obj.GetValue(SortGroupSummaryWithColumnProperty);
        }
        public static void SetSortGroupSummaryWithColumn(DependencyObject obj, bool value) {
            obj.SetValue(SortGroupSummaryWithColumnProperty, value);
        }
        public static readonly DependencyProperty SortGroupSummaryWithColumnProperty =
            DependencyProperty.RegisterAttached("SortGroupSummaryWithColumn", typeof(bool), typeof(GroupSummaryOptions), new PropertyMetadata(false));
        #endregion

        #region AlreadyHandled
        static bool GetAlreadyHandled(DependencyObject obj) {
            return (bool)obj.GetValue(AlreadyHandledProperty);
        }
        static void SetAlreadyHandled(DependencyObject obj, bool value) {
            obj.SetValue(AlreadyHandledProperty, value);
        }
        static readonly DependencyProperty AlreadyHandledProperty =
            DependencyProperty.RegisterAttached("AlreadyHandled", typeof(bool), typeof(GroupSummaryOptions), new PropertyMetadata(false));
        #endregion

        #endregion

        static GroupSummaryOptions() {
            EventManager.RegisterClassHandler(typeof(GridControl), GridControl.StartSortingEvent, new RoutedEventHandler(OnGridControlStartSorting));
        }

        static void OnGridControlStartSorting(object sender, RoutedEventArgs e) {
            GridControl Grid = (GridControl)sender;
            if (!GetSortGroupSummaryWithColumn(Grid) ||
                GetAlreadyHandled(Grid))
                return;
            SetAlreadyHandled(Grid, true);

            foreach (GridSortInfo sortInfo in Grid.SortInfo) {
                GridSummaryItem summaryItem = FindGridSummaryItem(Grid, sortInfo.FieldName);
                if (summaryItem != null) {
                    ArrayList groupedColumns = GetGroupedColumns(Grid);
                    ArrayList newSortInfoList = new ArrayList();
                    foreach (GridColumn col in groupedColumns) {
                        GridGroupSummarySortInfo groupSummaryInfo = FindGridGroupSummarySortInfo(Grid, summaryItem.FieldName);
                        if (groupSummaryInfo != null &&
                            groupSummaryInfo.SortOrder == sortInfo.SortOrder)
                            continue;

                        GridGroupSummarySortInfo newSortInfo = new GridGroupSummarySortInfo(summaryItem,
                                                                                            col.FieldName,
                                                                                            sortInfo.SortOrder);
                        newSortInfoList.Add(newSortInfo);
                        e.Handled = true;
                    }
                    Grid.GroupSummarySortInfo.AddRange((GridGroupSummarySortInfo[])newSortInfoList.ToArray(typeof(GridGroupSummarySortInfo)));
                    continue;
                }
            }

            SetAlreadyHandled(Grid, false);
        }

        static GridGroupSummarySortInfo FindGridGroupSummarySortInfo(GridControl Grid, string FieldName) {
            foreach (GridGroupSummarySortInfo sortInfo in Grid.GroupSummarySortInfo)
                if (sortInfo.SummaryItem.FieldName == FieldName)
                    return sortInfo;

            return null;
        }
        static GridSummaryItem FindGridSummaryItem(GridControl Grid, string FieldName) {
            foreach (GridSummaryItem SummaryItem in Grid.GroupSummary)
                if (SummaryItem.FieldName == FieldName)
                    return SummaryItem;

            return null;
        }

        static ArrayList GetGroupedColumns(GridControl Grid) {
            ArrayList result = new ArrayList();

            foreach (GridColumn col in Grid.Columns)
                if (col.GroupIndex > -1)
                    result.Add(col);

            result.Sort(new GridColumnGroupComparer());
            return result;
        }

        class GridColumnGroupComparer : IComparer {
            int IComparer.Compare(object x, object y) {
                return ((GridColumn)x).GroupIndex.CompareTo(((GridColumn)y).GroupIndex);
            }
        }
    }
}

File: VB/GroupSorting/GroupSummaryOptions.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows
Imports DevExpress.Xpf.Grid

Namespace GridControlOptions
    Public Class GroupSummaryOptions
#Region "Attached Properties"

#Region "SortGroupSummaryWithColumn"
        Public Shared Function GetSortGroupSummaryWithColumn(ByVal obj As DependencyObject) As Boolean
            Return CBool(obj.GetValue(SortGroupSummaryWithColumnProperty))
        End Function
        Public Shared Sub SetSortGroupSummaryWithColumn(ByVal obj As DependencyObject, ByVal value As Boolean)
            obj.SetValue(SortGroupSummaryWithColumnProperty, value)
        End Sub
        Public Shared ReadOnly SortGroupSummaryWithColumnProperty As DependencyProperty = DependencyProperty.RegisterAttached("SortGroupSummaryWithColumn", GetType(Boolean), GetType(GroupSummaryOptions), New PropertyMetadata(False))
#End Region

#Region "AlreadyHandled"
        Private Shared Function GetAlreadyHandled(ByVal obj As DependencyObject) As Boolean
            Return CBool(obj.GetValue(AlreadyHandledProperty))
        End Function
        Private Shared Sub SetAlreadyHandled(ByVal obj As DependencyObject, ByVal value As Boolean)
            obj.SetValue(AlreadyHandledProperty, value)
        End Sub
        Private Shared ReadOnly AlreadyHandledProperty As DependencyProperty = DependencyProperty.RegisterAttached("AlreadyHandled", GetType(Boolean), GetType(GroupSummaryOptions), New PropertyMetadata(False))
#End Region

#End Region

        Shared Sub New()
            EventManager.RegisterClassHandler(GetType(GridControl), GridControl.StartSortingEvent, New RoutedEventHandler(AddressOf OnGridControlStartSorting))
        End Sub

#Region "OnGridControlStartSorting"
        Private Shared Sub OnGridControlStartSorting(ByVal sender As Object, ByVal e As RoutedEventArgs)
            Dim Grid As GridControl = CType(sender, GridControl)
            If (Not GetSortGroupSummaryWithColumn(Grid)) OrElse GetAlreadyHandled(Grid) Then
                Return
            End If
            SetAlreadyHandled(Grid, True)

            For Each sortInfo As GridSortInfo In Grid.SortInfo
                Dim summaryItem As GridSummaryItem = FindGridSummaryItem(Grid, sortInfo.FieldName)
                If summaryItem IsNot Nothing Then
                    Dim groupedColumns As ArrayList = GetGroupedColumns(Grid)
                    Dim newSortInfoList As New ArrayList()
                    For Each col As GridColumn In groupedColumns
                        Dim groupSummaryInfo As GridGroupSummarySortInfo = FindGridGroupSummarySortInfo(Grid, summaryItem.FieldName)
                        If groupSummaryInfo IsNot Nothing AndAlso groupSummaryInfo.SortOrder = sortInfo.SortOrder Then
                            Continue For
                        End If

                        Dim newSortInfo As New GridGroupSummarySortInfo(summaryItem, col.FieldName, sortInfo.SortOrder)
                        newSortInfoList.Add(newSortInfo)
                        e.Handled = True
                    Next col
                    Grid.GroupSummarySortInfo.AddRange(CType(newSortInfoList.ToArray(GetType(GridGroupSummarySortInfo)), GridGroupSummarySortInfo()))
                    Continue For
                End If
            Next sortInfo

            SetAlreadyHandled(Grid, False)
        End Sub

        Private Shared Function FindGridGroupSummarySortInfo(ByVal Grid As GridControl, ByVal FieldName As String) As GridGroupSummarySortInfo
            For Each sortInfo As GridGroupSummarySortInfo In Grid.GroupSummarySortInfo
                If sortInfo.SummaryItem.FieldName = FieldName Then
                    Return sortInfo
                End If
            Next sortInfo

            Return Nothing
        End Function
        Private Shared Function FindGridSummaryItem(ByVal Grid As GridControl, ByVal FieldName As String) As GridSummaryItem
            For Each SummaryItem As GridSummaryItem In Grid.GroupSummary
                If SummaryItem.FieldName = FieldName Then
                    Return SummaryItem
                End If
            Next SummaryItem

            Return Nothing
        End Function

        Private Shared Function GetGroupedColumns(ByVal Grid As GridControl) As ArrayList
            Dim result As New ArrayList()

            For Each col As GridColumn In Grid.Columns
                If col.GroupIndex > -1 Then
                    result.Add(col)
                End If
            Next col

            result.Sort(New GridColumnGroupComparer())
            Return result
        End Function

        Private Class GridColumnGroupComparer
            Implements IComparer
            Private Function IComparer_Compare(ByVal x As Object, ByVal y As Object) As Integer Implements IComparer.Compare
                Return (CType(x, GridColumn)).GroupIndex.CompareTo((CType(y, GridColumn)).GroupIndex)
            End Function
        End Class
#End Region
    End Class
End Namespace
